File: Packages/OuraProtocol/Sources/OuraProtocol/HypnogramAssembler.swift
import Foundation

// SleepNet writes a completed night's phase codes to history in a compact burst after the night ends.
// The records' envelope ring timestamps therefore describe the write/finalization burst, not each
// sleep epoch. This platform-pure accumulator preserves the wire arrival order and reconstructs the
// sequence backward from a caller-supplied sleep-window end at the observed 30-second epoch cadence.

/// One reconstructed SleepNet epoch. `ts` is the Unix-second start of the 30-second stage interval.
public struct OuraHypnogramEpoch: Equatable, Sendable, Codable {
    public let stage: OuraSleepStage
    public let ts: Int

    public init(stage: OuraSleepStage, ts: Int) {
        self.stage = stage
        self.ts = ts
    }
}

/// One atomic sleep-phase record as fed to the burst assembler.
public struct OuraHypnogramRecord: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let stages: [OuraSleepStage]
    /// One flag per stage. Unwritten slots keep their position in the time axis but are not emitted.
    public let unwritten: [Bool]
    /// Optional resolved Unix time of the record envelope. It is observability/anchoring evidence only;
    /// phase order always follows arrival order rather than sorting these near-identical write times.
    public let envelopeUnixSeconds: Int?

    public init(ringTimestamp: UInt32, stages: [OuraSleepStage], unwritten: [Bool]? = nil,
                envelopeUnixSeconds: Int? = nil) {
        if let unwritten {
            precondition(unwritten.count == stages.count,
                         "Oura hypnogram unwritten flags must match the stage count")
        }
        self.ringTimestamp = ringTimestamp
        self.stages = stages
        self.unwritten = unwritten ?? Array(repeating: false, count: stages.count)
        self.envelopeUnixSeconds = envelopeUnixSeconds
    }
}

/// One completed burst of consecutive sleep-phase records, usually one whole-night hypnogram.
public struct OuraHypnogramBurst: Equatable, Sendable, Codable {
    /// Records remain in arrival/log order. Never sort them by their finalization timestamps.
    public let records: [OuraHypnogramRecord]

    public init(records: [OuraHypnogramRecord]) {
        self.records = records
    }

    public var totalCodes: Int { records.reduce(0) { $0 + $1.stages.count } }
    public var lastRingTimestamp: UInt32 { records.last?.ringTimestamp ?? 0 }

    /// The most recent usable envelope anchor in arrival order, if the caller resolved one.
    public var lastEnvelopeUnixSeconds: Int? {
        records.lazy.reversed().compactMap(\.envelopeUnixSeconds).first
    }

    /// Surfaces out-of-order envelope clocks without reordering the authoritative phase sequence.
    public var hasNonMonotonicRingTimes: Bool {
        zip(records, records.dropFirst()).contains { $1.ringTimestamp < $0.ringTimestamp }
    }

    /// Lay all stages backward from `endUnixSeconds`. For N stages, stage j starts at
    /// `end - (N - j) * secondsPerCode`, so the final interval ends exactly at `end`.
    ///
    /// If a 0x49 onset is supplied, leading epochs before it are clipped. A clamp that would remove
    /// every epoch is ignored so a stale or mismatched window cannot erase an otherwise valid burst.
    public func codesWithTimes(endUnixSeconds: Int,
                               sleepStartUnixSeconds: Int? = nil,
                               secondsPerCode: Int = 30) -> [OuraHypnogramEpoch] {
        guard secondsPerCode > 0, totalCodes > 0 else { return [] }
        let (duration, durationOverflow) = totalCodes.multipliedReportingOverflow(by: secondsPerCode)
        guard !durationOverflow else { return [] }
        let (firstTimestamp, timestampOverflow) = endUnixSeconds.subtractingReportingOverflow(duration)
        guard !timestampOverflow else { return [] }

        var output: [OuraHypnogramEpoch] = []
        output.reserveCapacity(totalCodes)
        var index = 0
        for record in records {
            for (recordIndex, stage) in record.stages.enumerated() {
                let (offset, offsetOverflow) = index.multipliedReportingOverflow(by: secondsPerCode)
                guard !offsetOverflow else { return [] }
                let (timestamp, additionOverflow) = firstTimestamp.addingReportingOverflow(offset)
                guard !additionOverflow else { return [] }
                if !record.unwritten[recordIndex] {
                    output.append(OuraHypnogramEpoch(stage: stage, ts: timestamp))
                }
                index += 1
            }
        }

        guard let sleepStartUnixSeconds else { return output }
        let clipped = output.filter { $0.ts >= sleepStartUnixSeconds }
        return clipped.isEmpty ? output : clipped
    }
}

/// Accumulates phase records whose finalization envelopes are within one burst gap. Oura ring ticks
/// are 100 ms, so the default 600-tick threshold allows a 60-second write burst while separating
/// distinct nights or analysis passes.
public final class OuraHypnogramAssembler {
    public let burstGapTicks: UInt32
    private var current: [OuraHypnogramRecord] = []

    /// Ring timestamps use 100 ms ticks. Keep the independently anchored wall-clock boundary in step
    /// with the configured raw-tick boundary, rounding a partial second upward.
    private var burstGapSeconds: Int {
        Int(burstGapTicks / 10 + (burstGapTicks % 10 == 0 ? 0 : 1))
    }

    public init(burstGapTicks: UInt32 = 600) {
        self.burstGapTicks = burstGapTicks
    }

    /// Feed a decoded atomic phase series. Returns the previous burst when this record starts a new
    /// one; otherwise returns nil while the current burst grows.
    public func feed(_ series: OuraSleepPhaseSeries,
                     envelopeUnixSeconds: Int? = nil) -> OuraHypnogramBurst? {
        feed(record: OuraHypnogramRecord(ringTimestamp: series.ringTimestamp,
                                         stages: series.stages,
                                         unwritten: series.unwritten,
                                         envelopeUnixSeconds: envelopeUnixSeconds))
    }

    /// Convenience for replay paths that already hold the record fields separately.
    public func feed(ringTimestamp: UInt32, stages: [OuraSleepStage], unwritten: [Bool]? = nil,
                     envelopeUnixSeconds: Int? = nil) -> OuraHypnogramBurst? {
        feed(record: OuraHypnogramRecord(ringTimestamp: ringTimestamp,
                                         stages: stages,
                                         unwritten: unwritten,
                                         envelopeUnixSeconds: envelopeUnixSeconds))
    }

    /// Feed one value record. Empty records are ignored and never open a burst.
    public func feed(record: OuraHypnogramRecord) -> OuraHypnogramBurst? {
        guard !record.stages.isEmpty else { return nil }
        if let previous = current.last {
            let gap = record.ringTimestamp >= previous.ringTimestamp
                ? record.ringTimestamp - previous.ringTimestamp
                : previous.ringTimestamp - record.ringTimestamp
            let crossesEnvelopeBoundary: Bool
            switch (previous.envelopeUnixSeconds, record.envelopeUnixSeconds) {
            case let (previousUnix?, currentUnix?):
                let wallGap = currentUnix >= previousUnix
                    ? currentUnix - previousUnix
                    : previousUnix - currentUnix
                crossesEnvelopeBoundary = wallGap > burstGapSeconds
            case (nil, nil):
                crossesEnvelopeBoundary = false
            default:
                // Mixed anchor availability is ambiguous at an archive/clock boundary. Splitting may
                // withhold a partial burst, but can never let unanchored stale slots complete a new night.
                crossesEnvelopeBoundary = true
            }
            if gap > burstGapTicks || crossesEnvelopeBoundary {
                let completed = OuraHypnogramBurst(records: current)
                current = [record]
                return completed
            }
        }
        current.append(record)
        return nil
    }

    /// Close and return the in-progress burst at history-drain end, or nil if no records are pending.
    public func finish() -> OuraHypnogramBurst? {
        guard !current.isEmpty else { return nil }
        let completed = OuraHypnogramBurst(records: current)
        current.removeAll(keepingCapacity: true)
        return completed
    }

    /// Compatibility spelling for callers that treat drain completion as a flush.
    public func flush() -> OuraHypnogramBurst? { finish() }

    public func reset() {
        current.removeAll(keepingCapacity: true)
    }

    public var pendingRecordCount: Int { current.count }
}

File: android/app/src/main/java/com/noop/oura/HypnogramAssembler.kt
package com.noop.oura

/** SleepNet classifies one stage per 30-second epoch. */
const val OURA_SLEEPNET_EPOCH_SECONDS: Long = 30

/**
 * One decoded sleep-phase record in transport arrival order. [envelopeUnixSeconds] is optional
 * because UTC may not be resolvable until the surrounding history page establishes a qualified
 * ring-clock anchor. It names the record write/finalization time, not the time of its first stage.
 */
data class OuraHypnogramRecord(
    val ringTimestamp: Long,
    val stages: List<OuraSleepStage>,
    val envelopeUnixSeconds: Long? = null,
    /** True slots reserve time but are omitted from the emitted staged hypnogram. */
    val unwritten: List<Boolean> = List(stages.size) { false },
) {
    init {
        require(unwritten.size == stages.size) { "unwritten flags must match stage count" }
    }
}

/** One reconstructed SleepNet epoch; [startUnixSeconds] names the interval start. */
data class OuraHypnogramEpoch(
    val stage: OuraSleepStage,
    val startUnixSeconds: Long,
)

/** Consecutive sleep-phase records written during one SleepNet finalization burst. */
data class OuraHypnogramBurst(val records: List<OuraHypnogramRecord>) {
    /** Total 2-bit stage codes across the burst. */
    val totalCodes: Int get() = records.sumOf { it.stages.size }

    /** Last envelope ring timestamp in arrival order. */
    val lastRingTimestamp: Long get() = records.lastOrNull()?.ringTimestamp ?: 0L

    /** Last UTC envelope anchor supplied by the caller, if any record could be anchored. */
    val lastEnvelopeUnixSeconds: Long?
        get() = records.asReversed().firstNotNullOfOrNull { it.envelopeUnixSeconds }

    /**
     * Surfaces an envelope timestamp step backwards without reordering the codes. Envelope times
     * describe near-simultaneous writes, so arrival order remains the only sequence ground truth.
     */
    val hasNonMonotonicRingTimes: Boolean
        get() = records.zipWithNext().any { (previous, next) ->
            next.ringTimestamp < previous.ringTimestamp
        }

    /**
     * Lay all stage codes backward from [endUnixSeconds]. Code j of N starts at
     * `end - (N - j) * secondsPerCode`, so the final code ends exactly at the supplied sleep end.
     * Records and codes are never sorted; transport arrival order is preserved.
     *
     * When [sleepStartUnixSeconds] is present, leading epochs before that 0x49 onset are clipped. A
     * clamp that would remove the entire burst is ignored so a mis-paired window cannot erase a night.
     */
    fun codesWithTimes(
        endUnixSeconds: Long,
        sleepStartUnixSeconds: Long? = null,
        secondsPerCode: Long = OURA_SLEEPNET_EPOCH_SECONDS,
    ): List<OuraHypnogramEpoch> {
        require(secondsPerCode > 0) { "secondsPerCode must be positive" }
        val count = totalCodes
        val laidOut = ArrayList<OuraHypnogramEpoch>(count)
        var index = 0
        for (record in records) {
            for ((recordIndex, stage) in record.stages.withIndex()) {
                val timestamp = endUnixSeconds - (count - index) * secondsPerCode
                if (!record.unwritten[recordIndex]) {
                    laidOut.add(
                        OuraHypnogramEpoch(
                            stage = stage,
                            startUnixSeconds = timestamp,
                        ),
                    )
                }
                index += 1
            }
        }
        if (sleepStartUnixSeconds == null) return laidOut
        val clipped = laidOut.filter { it.startUnixSeconds >= sleepStartUnixSeconds }
        return if (clipped.isEmpty()) laidOut else clipped
    }
}

/**
 * Groups phase records by their envelope ring-time proximity. SleepNet writes a completed night as
 * one burst; records within [burstGapTicks] belong together only when their UTC anchor availability
 * agrees and any two anchored envelopes are also within the equivalent wall-clock gap. Ring time is
 * in 100-ms ticks, so the default 600-tick gap is 60 seconds.
 */
class OuraHypnogramAssembler(
    val burstGapTicks: Long = 600,
) {
    init {
        require(burstGapTicks >= 0) { "burstGapTicks must be non-negative" }
    }

    private var current = ArrayList<OuraHypnogramRecord>()

    /** Feed an already-decoded phase series. */
    fun feed(
        series: OuraSleepPhaseSeries,
        envelopeUnixSeconds: Long? = null,
    ): OuraHypnogramBurst? = feed(
        ringTimestamp = series.ringTimestamp,
        stages = series.stages,
        envelopeUnixSeconds = envelopeUnixSeconds,
        unwritten = series.unwritten,
    )

    /**
     * Feed one record. Returns the previous burst when this record starts a new burst, or null while
     * the current burst is still accumulating. Records with no stage codes are ignored.
     */
    fun feed(
        ringTimestamp: Long,
        stages: List<OuraSleepStage>,
        envelopeUnixSeconds: Long? = null,
        unwritten: List<Boolean> = List(stages.size) { false },
    ): OuraHypnogramBurst? {
        if (stages.isEmpty()) return null
        require(unwritten.size == stages.size) { "unwritten flags must match stage count" }
        val record = OuraHypnogramRecord(
            ringTimestamp = ringTimestamp,
            stages = stages.toList(),
            envelopeUnixSeconds = envelopeUnixSeconds,
            unwritten = unwritten.toList(),
        )
        val previous = current.lastOrNull()
        if (previous != null) {
            val ringGap = if (ringTimestamp >= previous.ringTimestamp) {
                ringTimestamp - previous.ringTimestamp
            } else {
                previous.ringTimestamp - ringTimestamp
            }
            val previousEnvelope = previous.envelopeUnixSeconds
            val anchorAvailabilityChanged =
                (previousEnvelope == null) != (envelopeUnixSeconds == null)
            val wallGapExceeded = if (previousEnvelope != null && envelopeUnixSeconds != null) {
                val wallGap = if (envelopeUnixSeconds >= previousEnvelope) {
                    envelopeUnixSeconds - previousEnvelope
                } else {
                    previousEnvelope - envelopeUnixSeconds
                }
                wallGap > burstGapTicks / 10
            } else {
                false
            }
            if (ringGap > burstGapTicks || anchorAvailabilityChanged || wallGapExceeded) {
                val completed = OuraHypnogramBurst(current.toList())
                current = arrayListOf(record)
                return completed
            }
        }
        current.add(record)
        return null
    }

    /** Close and return the in-progress burst, or null when there is no pending record. */
    fun flush(): OuraHypnogramBurst? {
        if (current.isEmpty()) return null
        val completed = OuraHypnogramBurst(current.toList())
        current = ArrayList()
        return completed
    }

    /** Discard partial state before beginning a fresh history session. */
    fun reset() {
        current = ArrayList()
    }

    /** Number of records currently accumulating; useful for bounded-state diagnostics. */
    val pendingRecordCount: Int get() = current.size
}
